cloud-content/impl/src/main/java/coza/opencollab/sakai/cloudcontent/BlobStoreFileSystemHandler.java

```diff
@@ -182,15 +234,73 @@ public void destroy() throws IOException{
      * {@inheritDoc}
      */
     @Override
+    public URI getAssetDirectLink(String id, String root, String filePath) throws IOException {
+        ContainerAndName can = getContainerAndName(id, root, filePath);
+        HttpRequest hr = context.getSigner().signGetBlob(can.container, can.name, SIGNED_URL_VALIDITY_SECONDS);
+        if (hr == null) {
+            throw new IOException("No object found to creat signed url " + id);
+        }
+
+        return hr.getEndpoint();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
     public InputStream getInputStream(String id, String root, String filePath) throws IOException {
         ContainerAndName can = getContainerAndName(id, root, filePath);
         Blob blob = getBlobStore().getBlob(can.container, can.name);
         if (blob == null){
             throw new IOException("No object found for " + id);
         }
-        //we copy this to a byte array first since Sakai does some funny stuff 
-        //and with the stream and then swift and sakai don't play nice.
-        return new ByteArrayInputStream(FileCopyUtils.copyToByteArray(blob.getPayload().openStream()));
+
+        StorageMetadata metadata = blob.getMetadata();
+        Long size = metadata.getSize();
+
+        if (size != null && size.longValue() > maxBlobStreamSize) {
+            return streamFromTempFile(blob, size);
+        } else {
+            return blob.getPayload().openStream();
+        }
+    }
+
+    // Hacky implementation of downloading Blobs to temp files...
+    // This should probably happen in a specified location and use
+    // hashing to be sure of contents before reusing.
+    private InputStream streamFromTempFile(Blob blob, Long filesize) {
+        StorageMetadata metadata = blob.getMetadata();
+        String name = metadata.getName();
+        String filename = name + "-" + filesize;
+        filename = DigestUtils.md5Hex(filename);
+        FileInputStream stream = null;
+
+        // See if the temp file already exists
+        File check;
+        if (temporaryBlobDirectory != null) {
+            check = new File(temporaryBlobDirectory, filename);
+        }
+        else {
+            check = new File(System.getProperty("java.io.tmpdir"), filename);
+        }
+
+        if (check.exists()) {
+            try {
+                stream = new FileInputStream(check);
```

Comment:
Except we're trying to send an open stream here ..... we don't want this stream auto-closed